File: app/src/main/java/com/yawar/memo/modelView/CallHistoryModelView.java
package com.yawar.memo.modelView;

import android.annotation.SuppressLint;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.yawar.memo.model.CallModel;
import com.yawar.memo.retrofit.RetrofitClient;
import com.yawar.memo.utils.BaseApp;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class CallHistoryModelView extends ViewModel {

    BaseApp baseApp = BaseApp.getInstance();



    public MutableLiveData<ArrayList<CallModel>> callModelListMutableLiveData;
    public final ArrayList<CallModel> callModelsList ;
    SimpleDateFormat format;
    Date date;
    public MutableLiveData<Boolean> loading;

    @SuppressLint("SimpleDateFormat")
    public CallHistoryModelView() {
        callModelListMutableLiveData = new MutableLiveData<>();
        callModelsList = new ArrayList<>();
        format = new SimpleDateFormat("mm:ss");
        loading = new MutableLiveData<>(false);

    }

    @SuppressLint("CheckResult")
    public MutableLiveData<ArrayList<CallModel>> loadData(String my_id) {
        System.out.println("loadDataCall"+my_id);
        loading.postValue(true);

        Single<String> observable = RetrofitClient.getInstance().getapi().getMyCalls(my_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
        observable.subscribe(s->{
            loading.setValue(false);

            System.out.println("loadDataCall"+s);

            JSONArray jsonArray = null;
            try {
                 jsonArray = new JSONArray(s);

                for (int i = 0; i <= jsonArray.length() - 1; i++) {
                    JSONObject jsonObject = jsonArray.getJSONObject(i);
                    String userName = jsonObject.getString("first_name")+" "+jsonObject.getString("last_name");
                    String id = jsonObject.getString("id");
                    String caller_id = jsonObject.getString("caller");
                    String call_type = jsonObject.getString("call_type");
                    String call_status = jsonObject.getString("call_state");

                    String image = jsonObject.getString("profile_image");
                    String answer_id = jsonObject.getString("answer");
                    String duration = jsonObject.getString("duration");
                    String createdAt = jsonObject.getString("call_time");




                    callModelsList.add(new CallModel(id, userName, caller_id, image,call_type, answer_id, call_status,duration,createdAt));



                    }

                callModelListMutableLiveData.setValue(callModelsList);



            } catch (JSONException e) {
                e.printStackTrace();
                loading.setValue(false);

                callModelListMutableLiveData.setValue(null);
            }


        } ,s -> {
            loading.postValue(false);

            System.out.println("loadDataCallerrorr"+s);
            callModelListMutableLiveData.setValue(null);
        });

      return callModelListMutableLiveData;
    }

    Date getDateFromString(String dateString){

        try {
             date = format.parse(dateString);
            System.out.println(date+"dateeeeee");
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return date;
    }

}

File: app/src/main/java/com/yawar/memo/modelView/CallHistoryModelView.kt
package com.yawar.memo.modelView

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.yawar.memo.Api.GdgApi
import com.yawar.memo.constant.AllConstants
import com.yawar.memo.model.CallModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.log

class CallHistoryModelView : ViewModel() {

    var date: Date? = null


    private val _callModelListMutableLiveData = MutableLiveData<ArrayList<CallModel?>?>()
    val callModelListMutableLiveData: LiveData<ArrayList<CallModel?>?>
        get() = _callModelListMutableLiveData

    private val _loadingMutableLiveData =  MutableLiveData<Boolean>(false)
    val loadingMutableLiveData : LiveData<Boolean>
        get() = _loadingMutableLiveData
    var format: SimpleDateFormat = SimpleDateFormat("mm:ss")


    private var viewModelJob = Job()
    private val coroutineScope = CoroutineScope(viewModelJob + Dispatchers.Main )


    fun loadData( my_id : String): MutableLiveData<ArrayList<CallModel?>?>  {
        _loadingMutableLiveData.value = true
        val callModelsList =  ArrayList<CallModel?>()

        coroutineScope.launch {

//            val getResponeDeferred =GdgApi(AllConstants.base_node_url).apiService
//                .getMyCalls(my_id)
            val getResponeDeferred =GdgApi.apiService
                .getMyCalls(my_id)
            try {
                val listResult = getResponeDeferred?.await()
                _loadingMutableLiveData.value = false
                Log.d("ViewModel",listResult.toString() )

                val jsonArray = JSONArray(listResult)

                for (i in 0 until jsonArray.length()) {
                    val jsonObject: JSONObject = jsonArray.getJSONObject(i)
                    val userName =
                        jsonObject.getString("first_name") + " " + jsonObject.getString("last_name")
                    val id = jsonObject.getString("id")
                    val caller_id = jsonObject.getString("caller")
                    val call_type = jsonObject.getString("call_type")
                    val call_status = jsonObject.getString("call_state")
                    val image = jsonObject.getString("profile_image")
                    val answer_id = jsonObject.getString("answer")
                    val duration = jsonObject.getString("duration")
                    val createdAt = jsonObject.getString("call_time")
                    callModelsList.add(
                        CallModel(
                            id,
                            userName,
                            caller_id,
                            image,
                            call_type,
                            answer_id,
                            call_status,
                            duration,
                            createdAt
                        )
                    )
                }

                _callModelListMutableLiveData.value = callModelsList


            } catch (e: Exception) {
                _loadingMutableLiveData.value = false


                Log.d("getMarsRealEstateProperties: ","Failure: ${e.message}")

            }
        }
        return _callModelListMutableLiveData

    }

    override fun onCleared() {
        viewModelJob.cancel()
        super.onCleared()
    }

}